Xcode 9, swiftpm, and support for both Swift 3 & 4

sharplet opened the pull request:
Hopefully should be pretty self-explanatory. This change updates our dependencies, adds a Package.swift file, and makes the code portable across Swift 3.1, 3.2 and 4.0, tested in both Xcode 8 and the Xcode 9 beta.

Note that when compiling via Xcode (i.e., Carthage) we'll be compiling in Swift 3 language mode, but when compiling via swiftpm we'll use the latest version supported by the compiler (i.e., either Swift 3.1 or Swift 4.0).

sharplet committed:
Update to Nimble v7.0.1

sharplet committed:
Update to Regex 1.0.0

sharplet committed:
Ignore Cartfile.resolved

As a library, downstream users will ultimately end up using the latest
versions compatible with our public version constraints, which will
frequently be newer than whatever version we pin to whenever we update
our resolved libraries. This change ensures that whenever CI runs, it
runs using the latest compatible versions of our dependencies, rather
what we pinned to.

sharplet committed:
Add support for swiftpm

sharplet committed:
Only declare SWIFT_VERSION at xcodeproj level

sharplet committed:
Add support for compiling in Swift 4 language mode

sharplet commented on Package.swift:
```diff
@@ -0,0 +1,19 @@
+  ],
+  dependencies: [
+    .Package(url: "https://github.com/Quick/Quick.git", majorVersion: 1),
+    .Package(url: "https://github.com/sharplet/Regex.git", versions: Version(1, 0, 0, prereleaseIdentifiers: ["beta"])..<Version(2, 0, 0)),
```
Once https://github.com/sharplet/Regex/pull/44 is merged and tagged, we should be able to drop the beta stuff here.

sharplet replied:
Updated to [Regex 1.0.0](https://github.com/sharplet/Regex/releases/tag/1.0.0).

sharplet commented on .Package.test.swift:
```diff
@@ -0,0 +1,20 @@
+  dependencies: [
+    .Package(url: "https://github.com/Quick/Quick.git", majorVersion: 1),
+    .Package(url: "https://github.com/Quick/Nimble.git", majorVersion: 7),
+    .Package(url: "https://github.com/sharplet/Regex.git", versions: Version(1, 0, 0, prereleaseIdentifiers: ["beta"])..<Version(2, 0, 0)),
```
Ditto https://github.com/outware/Scenarios/pull/35/files#r121532355

chamander commented on Rakefile:
```diff
@@ -36,7 +46,7 @@ task :clean do
 end
 
-task :default => ["test:specs", "test:demo"]
+task :default => ["test:package", "test:specs", "test:demo"]
```
With the introduction of testing building with the Swift PM here, I think it would be neat to have a separate rake task, perhaps just `rake test` so that developers can run _both_ the unit tests and 'Demo' tests together, but separate from testing building through the Swift PM.

Do you think that would be useful? Would you be cool to add that in?

sharplet replied:
Sure. What about `test:xcode`?

sharplet replied:
I ended up making a new `test` action that's the default, and adding `test:all` for CI to run.

chamander commented on Rakefile:
```diff
@@ -24,6 +24,16 @@ namespace :test do
+      cp ".Package.test.swift", "Package.swift"
+      sh "swift test"
+    ensure
+      sh "git checkout HEAD -- Package.swift"
```
This is neat. 🙂 

Is the '--' here preferential? Up to this point, I have been checking out files from a particular tree-ish without, which would have this look like:

```
git checkout HEAD Package.swift
```

sharplet replied:
Most of the time it's equivalent, but there are times where git gets confused and can misinterpret a path as a branch or some other argument. The `--` is a way to unambiguously say that what follows should be treated as a path. Probably doesn't make a difference in this specific case.

chamander commented on Scenarios/StepDefinition.swift:
```diff
@@ -27,8 +27,9 @@ open class StepDefinition: QuickSpec {
       guard let (args, definition) = stepDefinitions.lazy
-        .flatMap({ pattern, function in
-          pattern.firstMatch(in: description).map {
+        .flatMap({ args -> (StepArguments, StepDefinitionFunc)? in
```
I think that the style here makes this a bit of an oddball. 

Would you be happy to change this to use a trailing closure?

sharplet replied:
I was never quite happy with how this turned out, so I did a bit of a refactor: a2ce62d.

chamander commented on Scenarios/Scenario.swift:
```diff
@@ -17,13 +17,22 @@ public final class Scenario: Preparable, Actionable, ScenarioBuilder {
     self.commitFunc = commit
   }
 
+  public convenience init(_ name: String, file: String = #file, line: UInt = #line) {
```
I like the looks of this separation. What was the motivation behind preferring the use of a convenience initialiser, over a default parameter? May I hazard a guess and say it's for autocompletion and visibility?

sharplet replied:
Unfortunately the default argument doesn't actually compile under Swift 4: turns out you can't use something that's `private` as a default value in something that's `public` (I think this is an ABI stability fix). Using a convenience initialiser means the implementation is private, and is free to reference private members.

chamander reviewed (commented):
This looks great. 

It builds, and tests pass, on my machine with Xcode 8.3.3 and the current beta of Xcode 9, when built through both Carthage, and the Swift Package Manager.

I've left a few points that I am curious about, and am keen to hear what you think / your rationale!

sharplet committed:
Add rake test action, differentiated from rake test:all

The default action is now `test`, which only runs the specs and UI tests
via Xcode, to avoid running the specs a second time via swiftpm.

CI now explicitly runs the new `test:all` action.

sharplet committed:
Extract named tuples to clarify how step definitions are matched

sharplet commented:
Unsatisfied with the duplication in `.Package.test.swift`, and inspired by [ReactiveSwift's package manifest](https://github.com/ReactiveCocoa/ReactiveSwift/blob/d6268d3167f900ed58b263738eb5c2425a4e97f1/Package.swift), I decided to go with a new approach where the dependency list is dynamic based on the presence of an environment variable. Take a look at 83732d7.

sharplet commented:
Should be good to go once we're green! 🍏

chamander commented on .travis.yml:
```diff
@@ -4,4 +4,4 @@ install:
   - gem install xcpretty
 before_script: carthage bootstrap --platform iphoneos
-script: rake
+script: rake test:all
```
Sweet! 🙂

chamander reviewed (commented):
Thanks for that change!

chamander commented on Scenarios/StepDefinition.swift:
```diff
@@ -61,7 +65,7 @@ open class StepDefinition: QuickSpec {
 
   private func registerStep(withPattern pattern: String, definition: @escaping StepDefinitionFunc) {
-    let step: (Regex, StepDefinitionFunc) = (regex(forPattern: pattern), definition)
+    let step = UnmatchedStep(pattern: regex(forPattern: pattern), function: definition)
```
Neat. I really like the way this has cleared up the tuple here.
